gtsfm/view_graph_estimator/reprojection_error_view_graph_estimator.py

```diff
@@ -0,0 +1,493 @@
+"""
+In the spirit of "Growing Consensus" (Son16cvpr)
+https://www.cv-foundation.org/openaccess/content_cvpr_2016/papers/Son_Solving_Small-Piece_Jigsaw_CVPR_2016_paper.pdf
+
+Authors: John Lambert
+"""
+
+import logging
+import os
+import time
+from dataclasses import dataclass
+from enum import Enum
+from types import SimpleNamespace
+from typing import Dict, List, Optional, Set, Tuple
+
+import dask
+import matplotlib.pyplot as plt
+import numpy as np
+from gtsam import Cal3Bundler, PinholeCameraCal3Bundler, Pose3, Rot3, Unit3
+
+
+import gtsfm.densify.mvs_utils as mvs_utils
+import gtsfm.multi_view_optimizer as multi_view_optimizer
+import gtsfm.utils.geometry_comparisons as comp_utils
+import gtsfm.utils.graph as graph_utils
+import gtsfm.utils.logger as logger_utils
+import gtsfm.utils.metrics as metrics_utils
+from gtsfm.averaging.rotation.shonan import ShonanRotationAveraging
+from gtsfm.averaging.translation.averaging_1dsfm import TranslationAveraging1DSFM
+from gtsfm.bundle.bundle_adjustment import BundleAdjustmentOptimizer
+from gtsfm.common.keypoints import Keypoints
+from gtsfm.data_association.data_assoc import DataAssociation
+from gtsfm.data_association.point3d_initializer import TriangulationOptions, TriangulationSamplingMode
+from gtsfm.evaluation.metrics import GtsfmMetricsGroup
+from gtsfm.two_view_estimator import TwoViewEstimationReport
+from gtsfm.view_graph_estimator.view_graph_estimator_base import ViewGraphEstimatorBase
+
+
+logger = logger_utils.get_logger()
+
+
+MIN_ALLOWED_TRI_ANGLE_DEG = 2
+MIN_ALLOWED_SUPPORT = 220
+# simulated (from global) vs. local measurement discrepancy
+MAX_ALLOWED_ROT_DISCREPANCY_DEG = 5
+MAX_ALLOWED_TRANS_DISCREPANCY_DEG = 5
+
+MAX_ALLOWED_ROT_CYCLE_ERROR_DEG = 5
+SUPPORT_CUTOFF_REPROJ_THRESHOLD_PX = 20
+
+
+@dataclass
+class TripletOptimizationReport:
+    """
+    """
+    triplet: Tuple[int,int,int]
+    is_success: bool = False
+    min_tri_angle: float = np.nan
+    support: int = np.nan
+    max_rot_discrepancy_deg: float = np.nan
+    max_trans_discrepancy_deg: float = np.nan
+    max_rot_error: float = np.nan
+    max_trans_error: float = np.nan
+
+    @property
+    def i0(self) -> int:
+        return triplet[0]
+
+    @property
+    def i1(self) -> int:
+        return triplet[1]
+
+    @property
+    def i2(self) -> int:
+        return triplet[2]
+
+
+class ThreeViewRegistrationMethod(str, Enum):
+    """Two supported modes for 3-view registration:
+    1. PNP (a la incremental SfM) or
+    2. Global rotation averaging + translation averaging.
+         w/ no BA, and immediately check compare pose differences (or reprojection errors).
```

Comment:
Would having 2 separate classes be better? They seem like a major algorithmic difference.